File: 50-powx-n/powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
         if (n == 0) return 1; // Base case: x^0 = 1
        
        long exp = n; // Convert int to long to avoid overflow
        if (exp < 0) {
            x = 1 / x;
            exp = -exp;
        }

        double result = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) result *= x; // If odd, multiply x
            x *= x; // Square x
            exp >>= 1; // Divide exponent by 2
        }

        return result;
        
    }
};

File: 50-powx-n/powx-n.java
class Solution {

    private double pow(double x, long n) {
        if (n == 0) return 1.0;

        double half = pow(x, n / 2);

        if (n % 2 == 0) {
            return half * half;      // even
        } else {
            return half * half * x;  // odd
        }
    }

    public double myPow(double x, int n) {
        long N = n;  // convert to long to handle Integer.MIN_VALUE

        if (N < 0) {
            x = 1 / x;
            N = -N;
        }

        return pow(x, N);
    }
}
